pkg/source/pubsub/pubsub_source.go

```diff
@@ -93,23 +113,53 @@ var ConfigPair = config.ConfigurationPair{
 }
 
 // newPubSubSource creates a new client for reading messages from PubSub
-func newPubSubSource(concurrentWrites int, projectID string, subscriptionID string) (*pubSubSource, error) {
+func newPubSubSource(concurrentWrites int, projectID string, subscriptionID string, maxOutstandingMessages, maxOutstandingBytes int, minExtensionPeriodSeconds int, streamingPullGoRoutines int, grpcConnectionPool int) (*pubSubSource, error) {
 	ctx := context.Background()
 
 	// Ensures as even as possible distribution of UUIDs
 	uuid.EnableRandPool()
 
-	client, err := pubsub.NewClient(ctx, projectID)
+	log := log.WithFields(log.Fields{"source": "pubsub", "cloud": "GCP", "project": projectID, "subscription": subscriptionID})
+
+	// We use a slice to provide the grpcConnectionPool option only if it is set.
+	// Otherwise we'll overwrite the client's clever under-the-hood default behaviour:
+	// https://github.com/googleapis/google-cloud-go/blob/380e7d23e69b22ab46cc6e3be58902accee2f26a/pubsub/pubsub.go#L165-L177
+	var opt []option.ClientOption
+	if grpcConnectionPool != 0 {
+		opt = append(opt, option.WithGRPCConnectionPool(grpcConnectionPool))
+	}
+
+	client, err := pubsub.NewClient(ctx, projectID, opt...)
 	if err != nil {
 		return nil, errors.Wrap(err, "Failed to create PubSub client")
 	}
 
+	// This temproary logic allows us to fix suboptimal behaviour without a breaking release.
+	// The order of priority is streaming_pull_goroutines > concurrent_writes > previous default
+	// We don't change the default becuase this would cause a major behaviour change in a non-major version bump
+
+	// If streamingPullGoRoutines is not set but concurrentWrites is, use concurrentWrites.
+	if streamingPullGoRoutines == 0 && concurrentWrites != 0 {
+		streamingPullGoRoutines = concurrentWrites
+		log.Warn("For the pubsub source, concurrent_writes is deprecated, and will be removed in the next major version. Use streaming_pull_goroutines instead")
+	}
+	// If neither are set, set it to the new defult, but warn users of this behaviour change
+	if streamingPullGoRoutines == 0 && concurrentWrites == 0 {
+		streamingPullGoRoutines = 50
+		log.Warn("Neither streaming_pull_goroutines nor concurrent_writes are set. The previous default is preserved, but strongly advise manual configuration of streaming_pull_goroutines, max_outstanding_messages and max_outstanding_bytes")
+	}
+	// Otherwise, streamingPullGoRoutines is set in the config and that value will be used.
+
 	return &pubSubSource{
-		projectID:        projectID,
-		client:           client,
-		subscriptionID:   subscriptionID,
-		concurrentWrites: concurrentWrites,
-		log:              log.WithFields(log.Fields{"source": "pubsub", "cloud": "GCP", "project": projectID, "subscription": subscriptionID}),
+		projectID:                 projectID,
+		client:                    client,
+		subscriptionID:            subscriptionID,
+		concurrentWrites:          concurrentWrites,
```

Comment:
nit: is the concurrent writes settings needed still here?

Reply:
It's not! Good spot, thank you.